src/compiler/checker.ts

```diff
@@ -8678,15 +8678,28 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 else {
                     context.tracker.trackSymbol(sym, context.enclosingDeclaration, meaning);
                 }
-                if (isIdentifier(node)) {
-                    const type = getDeclaredTypeOfSymbol(sym);
-                    const name = sym.flags & SymbolFlags.TypeParameter ? typeParameterToName(type, context) : factory.cloneNode(node);
-                    name.symbol = sym; // for quickinfo, which uses identifier symbol information
-                    return { introducesError, node: setTextRange(setEmitFlags(setOriginalNode(name, node), EmitFlags.NoAsciiEscaping), node) };
-                }
+                return { introducesError, node: attachSymbolToLeftmostIdentifier(node) as T };
             }
 
             return { introducesError, node };
+
+            /**
+             * Attaches a `.symbol` member to an identifier, cloning it to do so, so symbol information
+             * is smuggled out for symbol display information.
+             */
+            function attachSymbolToLeftmostIdentifier(node: Node): Node {

```

Comment:
Why not `<T extends Node>` here to avoid the cast above?

Reply:
Because that needs an `as T` cast in the `node === leftmost` branch anyway, so I'm just choosing which place of the two I'd rather cast at.

Reply:
Ah, yeah.